atlasdb-cassandra/src/main/java/com/palantir/atlasdb/cassandra/ReloadingCqlClusterContainer.java

```diff
@@ -0,0 +1,134 @@
+/*
+ * (c) Copyright 2022 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.atlasdb.cassandra;
+
+import com.palantir.atlasdb.cassandra.CassandraServersConfigs.CassandraServersConfig;
+import com.palantir.atlasdb.cassandra.backup.CqlCluster;
+import com.palantir.atlasdb.keyvalue.cassandra.async.client.creation.ClusterFactory.CassandraClusterConfig;
+import com.palantir.atlasdb.timelock.api.Namespace;
+import com.palantir.logsafe.logger.SafeLogger;
+import com.palantir.logsafe.logger.SafeLoggerFactory;
+import com.palantir.refreshable.Refreshable;
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.Optional;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Supplier;
+
+/**
+ * This container creates new CQL Clusters when the underlying server refreshable changes.
+ * After a refresh, previous CQL Clusters are closed to avoid resource leaks.
+ *
+ * There is no guarantee that a cluster provided via {@link #get} will <i>not</i> be closed whilst in use.
+ * Consumers should be resilient to any uses after the cluster has been closed, and where necessary, retry.
+ *
+ * After {@link #close()} has returned, no further CQL Clusters will be created, and all managed CQL Clusters will be
+ * closed.
+ */
+public final class ReloadingCqlClusterContainer implements Closeable, Supplier<CqlCluster> {
```

Comment:
This is nice, but:
1. possibly annoying to test because you've got to mess about with heavy-handed configs and CQL clusters
2. an awesome pattern that we _may_ want in the future

I think this could be nice if we make it parameterised and then could theoretically reuse it for other places where this is relevant (maybe none right now, but possibly in the future!)

Reply:
unrelated (and maybe you disagree), but I don't see what we gain from making this implement `Supplier<CqlCluster>` - we're depending upon a RCCC anyway, so we may as well not implement that, and maybe change the name of the method to reflect what we're really getting (e.g. `current` or `getCurrentOrThrow` or something)

Reply:
On Supplier<CqlCluster> - yeah it's a valid point. At some point in time I thought about maybe passing this around (and callees would only need to care that they can get a cluster, not that they can close the cluster container), but that isn't needed.

Reply:
As an aside, this container may need to become more sophisticated to handle in flight requests. I've already implemented that (and actually was the first design), but removed that guarantee to simplify the class massively.

Have an open discussion with the internal Cass experts on this! At least for restore, we don't need the guarantee, and I'll make any changes to the container in a separate PR.

Reply:
Sanity check: there are codepaths that use cql cluster that will continue to use it directly (without live reloading), and that will be changed in a future PR?

Reply:
There are no other code paths that use the CqlCluster in AtlasDB.